File: builder.py
class Outfit:
    def __init__(self):
        self.items = []

    def add(self, item):
        self.items.append(item)

    def get_total_cost(self):
        return sum(item.get_cost() for item in self.items)

    def __str__(self):
        text = "Outfit:\n"
        text += "\n".join(str(item) for item in self.items)
        text += f"\nTotal Cost: ${self.get_total_cost():.2f}"
        return text


class ClothingItem:
    def __init__(self, name, cost):
        self.name = name
        self.cost = cost

    def get_cost(self):
        return self.cost

    def __str__(self):
        return f"{self.name}: ${self.cost:.2f}"


class OutfitBuilder:
    def __init__(self):
        self.outfit = Outfit()

    def add_top(self):
        pass

    def add_bottom(self):
        pass

    def add_shoes(self):
        pass

    def add_accessory(self):
        pass

    def reset(self):
        self.outfit = Outfit()

    def get_outfit(self):
        return self.outfit


class CasualOutfitBuilder(OutfitBuilder):
    def add_top(self):
        self.outfit.add(ClothingItem("Camiseta de algodón", 15.99))

    def add_bottom(self):
        self.outfit.add(ClothingItem("Jeans ajustados", 39.99))

    def add_shoes(self):
        self.outfit.add(ClothingItem("Zapatillas deportivas", 49.99))

    def add_accessory(self):
        self.outfit.add(ClothingItem("Reloj casual", 25.99))


class OutfitDirector:
    def __init__(self, builder):
        self.builder = builder

    def construct_full_outfit(self):
        self.builder.reset()
        self.builder.add_top()
        self.builder.add_bottom()
        self.builder.add_shoes()
        self.builder.add_accessory()

    def construct_basic_outfit(self):
        self.builder.reset()
        self.builder.add_top()
        self.builder.add_bottom()


# Prueba del código
casual_builder = CasualOutfitBuilder()
outfit_director = OutfitDirector(casual_builder)
outfit_director.construct_full_outfit()
print(casual_builder.get_outfit())

formal_builder = CasualOutfitBuilder()
formal_director = OutfitDirector(formal_builder)
formal_director.construct_basic_outfit()
print(formal_builder.get_outfit())

File: java/builder.java
import java.util.ArrayList;
import java.util.List;

// Clase que representa un Outfit (conjunto de prendas)
class Outfit {
    private List<ClothingItem> items;

    public Outfit() {
        items = new ArrayList<>();
    }

    public void add(ClothingItem item) {
        items.add(item);
    }

    public double getTotalCost() {
        double total = 0;
        for (ClothingItem item : items) {
            total += item.getCost();
        }
        return total;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("Outfit:\n");
        for (ClothingItem item : items) {
            sb.append(item).append("\n");
        }
        sb.append(String.format("Total Cost: $%.2f", getTotalCost()));
        return sb.toString();
    }
}

// Clase que representa una prenda de ropa
class ClothingItem {
    private String name;
    private double cost;

    public ClothingItem(String name, double cost) {
        this.name = name;
        this.cost = cost;
    }

    public double getCost() {
        return cost;
    }

    @Override
    public String toString() {
        return String.format("%s: $%.2f", name, cost);
    }
}

// Clase abstracta Builder para construir un Outfit
abstract class OutfitBuilder {
    protected Outfit outfit;

    public OutfitBuilder() {
        outfit = new Outfit();
    }

    public abstract void addTop();
    public abstract void addBottom();
    public abstract void addShoes();
    public abstract void addAccessory();

    public void reset() {
        outfit = new Outfit();
    }

    public Outfit getOutfit() {
        return outfit;
    }
}

// Implementación concreta de un builder para un Outfit casual
class CasualOutfitBuilder extends OutfitBuilder {
    @Override
    public void addTop() {
        outfit.add(new ClothingItem("Camiseta de algodón", 15.99));
    }

    @Override
    public void addBottom() {
        outfit.add(new ClothingItem("Jeans ajustados", 39.99));
    }

    @Override
    public void addShoes() {
        outfit.add(new ClothingItem("Zapatillas deportivas", 49.99));
    }

    @Override
    public void addAccessory() {
        outfit.add(new ClothingItem("Reloj casual", 25.99));
    }
}

// Director que utiliza el builder para construir diferentes outfits
class OutfitDirector {
    private OutfitBuilder builder;

    public OutfitDirector(OutfitBuilder builder) {
        this.builder = builder;
    }

    // Construye un outfit completo: top, bottom, shoes y accessory.
    public void constructFullOutfit() {
        builder.reset();
        builder.addTop();
        builder.addBottom();
        builder.addShoes();
        builder.addAccessory();
    }

    // Construye un outfit básico: solo top y bottom.
    public void constructBasicOutfit() {
        builder.reset();
        builder.addTop();
        builder.addBottom();
    }
}

// Clase pública con método main para ejecutar el código
public class builder {
    public static void main(String[] args) {
        // Construcción de un outfit completo (casual)
        OutfitBuilder casualBuilder = new CasualOutfitBuilder();
        OutfitDirector director = new OutfitDirector(casualBuilder);
        director.constructFullOutfit();
        System.out.println(casualBuilder.getOutfit());

        System.out.println(); // Línea en blanco para separar salidas

        // Construcción de un outfit básico (casual)
        OutfitBuilder basicBuilder = new CasualOutfitBuilder();
        OutfitDirector basicDirector = new OutfitDirector(basicBuilder);
        basicDirector.constructBasicOutfit();
        System.out.println(basicBuilder.getOutfit());
    }
}
